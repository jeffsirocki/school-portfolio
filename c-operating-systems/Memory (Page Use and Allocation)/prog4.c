// Tyler Bennett and Jeff Sirocki
// CS 3013 - Project 4
// Due: 02-27-15

// Implementing Virtual Memory
// Checkpoint Due: 02-23-2015

#define _BSD_SOURCE
#include <pthread.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <semaphore.h> 
#include <time.h>
#include "prog4.h"

// Threads
pthread_t user_threads[NUM_USER];

// User functions that the threads will run (TESTING PURPOSES)

// MEMORY MAXIMIZER ( SLOW )
void * run_user1(User *uptr){
	pthread_t U = uptr->thread;
	int uid = (int) uptr->id;
	int page;
	int job_num = 0;
	
	// LOOP until hit Max Page count
	while( TRUE && index_count < MAX) {
		
		// Increment Count
		sem_wait(&count);
		page = index_count;
		index_count++;
		sem_post(&count);
		job_num++;
		printf("USER %d on Job %d working on Page %d: ", uid, job_num, page);
		
		// User Functions
		indexes[page] = allocateNewInt(); 
		int *value = accessIntPtr(indexes[page]); 
		
		*value = (page * 3);
		
		unlockMemory(indexes[page]); 
	}
}

// MEMORY CACHER ( FAST )
void * run_user2(User *uptr){
	pthread_t U = uptr->thread;
	int uid = (int) uptr->id;
	int page;
	int job_num = 0;

	//printf("USER: \n");
	
	while( TRUE && index_count < MAX) {
		
		// Increment Count
		sem_wait(&count);
		index_count++;
		sem_post(&count);
		// Assign Page
		page = (rand() % 24);
		job_num++;
		printf("USER %d on Job %d working on Page %d: ", uid, job_num, page);
		if( index_count < 25 ) {
			indexes[page] = allocateNewInt(); 
		}
		
		// User Functions
		int *value = accessIntPtr(indexes[page]); 
		
		*value = (page * 3);
		
		unlockMemory(indexes[page]); 
		usleep(25000);
	}
}

// MEMORY TESTER ( Medium )
/* void * run_user3(User *uptr){
	
	// Declarations
	pthread_t U = uptr->thread;
	int uid = (int) uptr->id;
	int page, l;
	int job_num = 0;
	
	while( TRUE && index_count < MAX) {
		
		// Increment Count
		sem_wait(&count);
		page = index_count;
		index_count++;
		sem_post(&count);
		job_num++;
		
		// Allocate New Int
		if( index_count > 25 ) {
			
			page = (rand() % 50); // Number 0 - 24
			printf("USER %d on Job %d working on Page %d: ", uid, job_num, page);
			if ( page > 24 ) {
				indexes[page] = allocateNewInt(); 
			}
			
		} else if( index_count > 125 ) {
			
			page = (rand() % 125); // Number 25 to 125
			printf("USER %d on Job %d working on Page %d: ", uid, job_num, page);
		} else {
			indexes[page] = allocateNewInt(); 
		}
		
		// User Functions
		int *value = accessIntPtr(indexes[page]); 
		
		*value = (page * 3);
		
		unlockMemory(indexes[page]); 
		usleep(250);
	}
} */

// Main Function: simulates user-end
int main(){

	// Delcarations
	int i;
	User user[NUM_USER];
	struct timeval start;
	struct timeval end;
	double dstart, dend, dtotal;
	
	

	// Initializations
	index_count = 0;
	
	// Semaphores
	sem_init(&r, 0, 1);
	sem_init(&s, 0, 1);
	sem_init(&h, 0, 1);
	sem_init(&rtype_counter, 0, 1);
	sem_init(&stype_counter, 0, 1);
	sem_init(&zlock, 0, 1);
	sem_init(&count, 0, 1);
	sem_init(&find, 0, 1);
	for( i = 0; i < 1000; i++) {
		sem_init(&pgTable[i], 0, 1);
		sem_init(&lock[i], 0, 1);
	}
	
	// Set start of simulation
	gettimeofday(&start, NULL);
    dstart = (start.tv_sec*1000.0) + (start.tv_usec/1000.0);
	
	// Create Threads
	for( i = 0; i < NUM_USER; i++) {
		user[i].id = i;
		user[i].thread = user_threads[i];
		if(pthread_create(&user_threads[i], NULL, run_user1, &user[i])){
			printf("ERROR: Unable to create thread.\n");
			return 1;
		}
	}

	// Join Threads
	for( i = 0; i < NUM_USER; i++) {
		if (pthread_join(user_threads[i], NULL)) { 
			printf("Error joining thread #%d!\n", 1);
		} 
	}
	
	// Free Memory
	for( i = 0; i < MAX; i++ ) {
		if ( indexes[i] != NULL ) {
			freeMemory(indexes[i]);
		}
	}
		
	// Set end of simulation
	gettimeofday(&end, NULL);
    dend = (end.tv_sec*1000.0) + (end.tv_usec/1000.0);
		
	// OUTPUT: Timing metrics
	dtotal = (dend - dstart)/1000.0;
	printf("Total Simulation took %lf seconds\n\n", dtotal);	
	
	return 0;
}