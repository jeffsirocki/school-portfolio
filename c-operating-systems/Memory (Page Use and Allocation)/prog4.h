// Tyler Bennett and Jeff Sirocki
// CS 3013 - Project 4
// Due: 02-27-15

// Implementing Virtual Memory
// Checkpoint Due: 02-23-2015

#include <semaphore.h>

#define DEBUG 1
#define TRUE 1
#define NUM_USER 10
#define MAX 150

// Memory Hierarchy
int RAM[25];
int SSD[100];
int HDD[1000];

//basically these new arrays are full of 0/1, 
// 0 indicates that nothing is currently in that location,
// 1 indicates something is there
int RAM_allocation[25];
int SSD_allocation[100];
int HDD_allocation[1000];

typedef unsigned short vAddr;

vAddr indexes[1000]; 
int index_count;
int z;

// Semaphores
sem_t pgTable[1000];
sem_t lock[1000];
sem_t r;
sem_t s;
sem_t h;
sem_t zlock;
sem_t rtype_counter;
sem_t stype_counter;
sem_t find;
sem_t count;

// Page Table
typedef struct PageTag { 
	int locked; // 0 if page unlocked, 1 if locked
	int allocated; // 0 if unallocated, 1 if allocated
	struct timeval time; // Info on when it was last accessed
	int type; // 1 RAM, 2 SSD, 3 HDD
	int index; // index into the allocation arrays 0<x<1000
} Page;

typedef struct UserTag {
	pthread_t thread;
	int id;
} User;


// Function Prototypes
vAddr allocateNewInt();
int *accessIntPtr(vAddr address);
void unlockMemory(vAddr address);
void freeMemory(vAddr address);
