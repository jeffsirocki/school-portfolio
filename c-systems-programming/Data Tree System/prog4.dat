 4  4  0  3 L7   6 C8 100 T4  19
 4  6  0  8 Z0  42 A1  14 W8   1 C4  65 B9  91 F6  12 F1 123 X8  18
 6  4  0  1 A3  43
10  4  0  4 L7  12 A1 234 G4  11 F6   4
10  4  1  6 L7   3 X8  23 B4  42 F6  20 C5  55 M6  71
12 12  1  5 D2 345 D1 245 D0 145 C9   9 L5   5
 6 12  0  2 A1  21 Y2 102
 4  4  0 12 L7   7 L6   6 L5   5 N5 456 O4 741 P3 013 Q2 871 R2 152 S1 111 T0 101 U2   1 V2  28
10  4  0  3 F6  80 G4  34 L7   8
 4 12  1  8 Z0  41 X8  22 Y7   7 N2  16 N1  15 A1  87 L7   1 F6  21
12  4  0  4 L6   6 X1  75 X2  65 X8   8
 6 12  1  3 L7   5 A1  18 X8  22
12 10  1  8 B8   2 D8   4 F6   5 X8   6 G4  24 K6  16 L7 900 S1   2
10  4  1 11 M6  17 N6 389 X8  34 Z0 123 D1  56 K9  59 L7  26 S8  77 S6  66 S5  41 T0  99
12 12  0  3 D2   3 E8   8 X8  55
 4  6  0  5 W8   4 X8  44 F6  12 B4  45 B9 182
10 12  1  2 A1   9 X4  33
12  6  1  3 L7   7 X8 910 F6   2
12 10  0  4 Z0  09 Z9  03 C9   7 Y7  44
10  4  1  6 L7   2 M7 007 F6  17 E4  48 M6  27 T0  1
 4  6  1  4 Z0  53 G4   2 A1   3 G4   3
12  4  0  3 L7   6 C8  11 T4  39
10  4  0  4 F6   1 C2   3 T0   4 L7 432
10 12  1  5 X4   3 A2 222 U2   2 V8  87 N5  72
12 12  1  5 L5  10 C5   4 D0  09 R7  73 P3  34
16 16 16

