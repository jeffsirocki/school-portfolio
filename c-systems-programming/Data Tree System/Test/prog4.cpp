// Jeff Sirocki and Joseph Kaiser
// CS2303 A02

#include <iostream>
#include <string>
#include "prog4.h"
using namespace std;

int main( void ) {
  
  int i; // counter

  int xpos, ypos, zpos; // Store Coordinates
  int k_items; // number of unique items on line of input
  string ID;
  int count;

  Item I;
  TreeNode root;
  //TreeNode BST;

  // Initialize BINARY TREE

  // While input is not termination case
  while( xpos != 16 && ypos != 16 && zpos != 16 ) {

    // Prompt
    cout << "Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n " << endl;

    // End case for { 16, 16, 16 }
    cin >> xpos >> ypos >> zpos;
    if ( xpos == 16 && ypos == 16 && zpos == 16 ) {
      cout << xpos << ypos << zpos << endl;
      // end while
      break; 
    }

    cin >> k_items;

    for( i = 0; i < k_items; i++ ) {
      cin >> ID >> count;

      // Set Item
      I.ID = ID;
      I.count = count;
      I.xpos = xpos;
      I.ypos = ypos;
      I.zpos = zpos;

      //cout << I.ID << " " << I.count << " " << I.xpos << " " << I.ypos << " " << I.zpos << endl;


      //cout << " ROOT: " << root.item << " " << root.leftPtr << " " << root.rightPtr << endl;
      // Insert into tree
      root.insert( I );
    
      // Print it
      //cout << ID << " " << count << endl;
    }
  }

  cout << " Got to Print Tree " << endl << endl;
  cout << " *********** PRINTING TREE ***************** " << endl << endl;

  // Print Binary Tree
  root.print_tree( );



  // End of Main
  return 0;
}
