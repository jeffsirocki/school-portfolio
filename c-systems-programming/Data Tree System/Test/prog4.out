Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
Enter store( x, y, z ) k-itmes Item-1 count-1 ... Item-n count-n 
161616
 Got to Print Tree 

 *********** PRINTING TREE ***************** 

 ITEM A1:
      234 -- (10, 4, 0)
      87 -- (4, 12, 1)
      21 -- (6, 12, 0)
      18 -- (6, 12, 1)
      14 -- (4, 6, 0)
      9 -- (10, 12, 1)
      3 -- (4, 6, 1)


 ITEM A2:
      222 -- (10, 12, 1)


 ITEM A3:
      43 -- (6, 4, 0)


 ITEM B4:
      45 -- (4, 6, 0)
      42 -- (10, 4, 1)


 ITEM B8:
      2 -- (12, 10, 1)


 ITEM B9:
      182 -- (4, 6, 0)
      91 -- (4, 6, 0)


 ITEM C2:
      3 -- (10, 4, 0)


 ITEM C4:
      65 -- (4, 6, 0)


 ITEM C5:
      55 -- (10, 4, 1)
      4 -- (12, 12, 1)


 ITEM C8:
      100 -- (4, 4, 0)
      11 -- (12, 4, 0)


 ITEM C9:
      9 -- (12, 12, 1)
      7 -- (12, 10, 0)


 ITEM D0:
      145 -- (12, 12, 1)
      9 -- (12, 12, 1)


 ITEM D1:
      245 -- (12, 12, 1)
      56 -- (10, 4, 1)


 ITEM D2:
      345 -- (12, 12, 1)
      3 -- (12, 12, 0)


 ITEM D8:
      4 -- (12, 10, 1)


 ITEM E4:
      48 -- (10, 4, 1)


 ITEM E8:
      8 -- (12, 12, 0)


 ITEM F1:
      123 -- (4, 6, 0)


 ITEM F6:
      80 -- (10, 4, 0)
      21 -- (4, 12, 1)
      20 -- (10, 4, 1)
      17 -- (10, 4, 1)
      12 -- (4, 6, 0)
      12 -- (4, 6, 0)
      5 -- (12, 10, 1)
      4 -- (10, 4, 0)
      2 -- (12, 6, 1)
      1 -- (10, 4, 0)


 ITEM G4:
      34 -- (10, 4, 0)
      24 -- (12, 10, 1)
      11 -- (10, 4, 0)
      3 -- (4, 6, 1)
      2 -- (4, 6, 1)


 ITEM K6:
      16 -- (12, 10, 1)


 ITEM K9:
      59 -- (10, 4, 1)


 ITEM L5:
      10 -- (12, 12, 1)
      5 -- (4, 4, 0)
      5 -- (12, 12, 1)


 ITEM L6:
      6 -- (12, 4, 0)
      6 -- (4, 4, 0)


 ITEM L7:
      900 -- (12, 10, 1)
      432 -- (10, 4, 0)
      26 -- (10, 4, 1)
      12 -- (10, 4, 0)
      8 -- (10, 4, 0)
      7 -- (12, 6, 1)
      7 -- (4, 4, 0)
      6 -- (12, 4, 0)
      6 -- (4, 4, 0)
      5 -- (6, 12, 1)
      3 -- (10, 4, 1)
      2 -- (10, 4, 1)
      1 -- (4, 12, 1)


 ITEM M6:
      71 -- (10, 4, 1)
      27 -- (10, 4, 1)
      17 -- (10, 4, 1)


 ITEM M7:
      7 -- (10, 4, 1)


 ITEM N1:
      15 -- (4, 12, 1)


 ITEM N2:
      16 -- (4, 12, 1)


 ITEM N5:
      456 -- (4, 4, 0)
      72 -- (10, 12, 1)


 ITEM N6:
      389 -- (10, 4, 1)


 ITEM O4:
      741 -- (4, 4, 0)


 ITEM P3:
      34 -- (12, 12, 1)
      13 -- (4, 4, 0)


 ITEM Q2:
      871 -- (4, 4, 0)


 ITEM R2:
      152 -- (4, 4, 0)


 ITEM R7:
      73 -- (12, 12, 1)


 ITEM S1:
      111 -- (4, 4, 0)
      2 -- (12, 10, 1)


 ITEM S5:
      41 -- (10, 4, 1)


 ITEM S6:
      66 -- (10, 4, 1)


 ITEM S8:
      77 -- (10, 4, 1)


 ITEM T0:
      101 -- (4, 4, 0)
      99 -- (10, 4, 1)
      4 -- (10, 4, 0)
      1 -- (10, 4, 1)


 ITEM T4:
      39 -- (12, 4, 0)
      19 -- (4, 4, 0)


 ITEM U2:
      2 -- (10, 12, 1)
      1 -- (4, 4, 0)


 ITEM V2:
      28 -- (4, 4, 0)


 ITEM V8:
      87 -- (10, 12, 1)


 ITEM W8:
      4 -- (4, 6, 0)
      1 -- (4, 6, 0)


 ITEM X1:
      75 -- (12, 4, 0)


 ITEM X2:
      65 -- (12, 4, 0)


 ITEM X4:
      33 -- (10, 12, 1)
      3 -- (10, 12, 1)


 ITEM X8:
      910 -- (12, 6, 1)
      55 -- (12, 12, 0)
      44 -- (4, 6, 0)
      34 -- (10, 4, 1)
      23 -- (10, 4, 1)
      22 -- (6, 12, 1)
      22 -- (4, 12, 1)
      18 -- (4, 6, 0)
      8 -- (12, 4, 0)
      6 -- (12, 10, 1)


 ITEM Y2:
      102 -- (6, 12, 0)


 ITEM Y7:
      44 -- (12, 10, 0)
      7 -- (4, 12, 1)


 ITEM Z0:
      123 -- (10, 4, 1)
      53 -- (4, 6, 1)
      42 -- (4, 6, 0)
      41 -- (4, 12, 1)
      9 -- (12, 10, 0)


 ITEM Z9:
      3 -- (12, 10, 0)


